clang/lib/Sema/SemaBounds.cpp

```diff
@@ -2327,8 +2518,50 @@ namespace {
         }
       }
 
+      // Update State.UEQ and State.G.
```

Comment:
I didn't see this comment addressed in your most recent change.  It would have been helpful to have a clarifying response here about why it wasn't addressed.

I looked into this and see that UpdateG is screening out non-modifying expressions in the S_i.   I am pretty concerned about this.  Expression equivalence checking assumes that expressions are non-modifying expressions and may produce incorrect results if they are modifying expressions.   

For this reason, I think we should have an invariant that none of the subexpressions G_i containy modifying expression. This change is already large and I'd rather land it now, then have it continune onward.  @kkjeer could you file an issue about this, and follow-up in a subsequent check-in.

Reply:
I created issue #815 to track this work.